apps/evm_family/evm_sign_msg.c

```diff
@@ -0,0 +1,436 @@
+  evm_send_result(&response);
+
+  return true;
+}
+
+static bool get_msg_data(evm_query_t *query) {
+  evm_result_t response = init_evm_result(EVM_RESULT_SIGN_MSG_TAG);
+  uint32_t total_size = sign_msg_ctx.init.total_msg_size;
+  const evm_sign_msg_data_t *msg_data = &query->sign_msg.msg_data;
+  const common_chunk_payload_t *payload = &(msg_data->chunk_payload);
+  const common_chunk_payload_chunk_t *chunk = &(payload->chunk);
+
+  uint32_t size = 0;
+
+  sign_msg_ctx.msg_data = malloc(sign_msg_ctx.init.total_msg_size);
+  ASSERT(NULL != sign_msg_ctx.msg_data);
+
+  while (1) {
+    // Get next data chunk from host
+    if (!evm_get_query(query, EVM_QUERY_SIGN_MSG_TAG) ||
+        !check_which_request(query, EVM_SIGN_MSG_REQUEST_MSG_DATA_TAG) ||
+        false == query->sign_msg.msg_data.has_chunk_payload) {
+      return false;
+    }
+
+    if (!msg_data->has_chunk_payload ||
+        payload->chunk_index >= payload->total_chunks ||
+        size + chunk->size > total_size) {
+      evm_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                     ERROR_DATA_FLOW_INVALID_DATA);
+      return false;
+    }
+
+    memcpy(sign_msg_ctx.msg_data + size, chunk->bytes, chunk->size);
+    size += chunk->size;
+
+    // Send chunk ack to host
+    response.sign_msg.which_response = EVM_SIGN_MSG_RESPONSE_DATA_ACCEPTED_TAG;
+    response.sign_msg.data_accepted.has_chunk_ack = true;
+    response.sign_msg.data_accepted.chunk_ack.chunk_index =
+        payload->chunk_index;
+    evm_send_result(&response);
+
+    // If no data remaining to be received from the host, then exit
+    if (0 == payload->remaining_size ||
+        payload->chunk_index + 1 == payload->total_chunks) {
+      break;
+    }
+  }
+  return true;
+}
+
+static bool get_user_verification() {
+  bool result = false;
+  switch (sign_msg_ctx.init.message_type) {
+    case EVM_SIGN_MSG_TYPE_ETH_SIGN: {
+      const size_t array_size = sign_msg_ctx.init.total_msg_size * 2 + 3;
+      char *buffer = malloc(array_size);
+      memzero(buffer, array_size);
+      snprintf(buffer, array_size, "0x");
+      byte_array_to_hex_string(sign_msg_ctx.msg_data,
+                               sign_msg_ctx.init.total_msg_size,
+                               buffer + 2,
+                               array_size - 2);
+      // TODO: Add a limit on size of data per confirmation based on LVGL buffer
+      // and split message into multiple confirmations accordingly
+      result = core_scroll_page(
+          UI_TEXT_VERIFY_MESSAGE, (const char *)buffer, evm_send_error);
+      memzero(buffer, array_size);
+      free(buffer);
+    } break;
+
+    case EVM_SIGN_MSG_TYPE_PERSONAL_SIGN: {
+      // TODO: Add a limit on size of data per confirmation based on LVGL buffer
+      // and split message into multiple confirmations accordingly
+      result = core_scroll_page(UI_TEXT_VERIFY_MESSAGE,
+                                (const char *)sign_msg_ctx.msg_data,
```

Comment:
verify if `sign_msg_ctx.msg_data` terminates with `NULL`

Reply:
updated

Reply:
Updated the solution and added extra character at the end of the data received